File: 백준/Gold/15683. 감시/감시.py
import sys
input = sys.stdin.readline

# 가능한 방향의 배열
mode = [
    [],
    [[0], [1], [2], [3]],
    [[0, 2], [1, 3]],
    [[0, 1], [1, 2], [2, 3], [0, 3]],
    [[0, 1, 2], [0, 1, 3], [1, 2, 3], [0, 2, 3]],
    [[0, 1, 2, 3]],
]

dx = [-1,0,1,0]
dy = [0,-1,0,1]

n, m = map(int, input().split())
graph = [list(map(int, input().split())) for _ in range(n)]

# cctv가 있는곳의 좌표 + cctv의 종류
cctv = []
for i in range(n):
    for j in range(m):
        if graph[i][j] != 0 and graph[i][j] != 6:
            cctv.append([i,j,graph[i][j]])

# cctv가 바라보고 있는 방향을 체크
def check(x,y,dir,tmp):
    for i in dir:
        nx, ny = x, y
        while True:
            nx += dx[i]
            ny += dy[i]
            # 범위 벗어나면 break
            if nx < 0 or nx >= n or ny < 0 or ny >= m:
                break
            # 벽을 만나면 break
            if tmp[nx][ny] == 6:
                break
            # 빈 공간이거나 이미 check한 곳 일 때
            elif tmp[nx][ny] == 0:
                # check표시
                tmp[nx][ny] = -1


ans = sys.maxsize
def dfs(depth, arr):
    global ans

    # 만약 모든 cctv의 경우를 다 체크했을 때
    if depth == len(cctv):
        cnt = 0
        for i in range(n):
            for j in range(m):
                if arr[i][j] == 0:
                    cnt += 1
        ans = min(ans, cnt)
        # 반환하여 다른 방향을 볼 때를 체크한다
        return
    
    # 백트래킹을 실시할 것이므로 깊은복사를 한 배열을 만들어준다
    tmp = [li1[:] for li1 in arr]
    x, y, num = cctv[depth]

    for i in mode[num]:
        check(x,y,i,tmp)
        dfs(depth+1, tmp)
        # check가 끝나고 다른 배열을 확인할 때 깊은복사를 한 배열 필요
        tmp = [li1[:] for li1 in arr]

dfs(0, graph)
print(ans)

File: 백준/Gold/15683. 감시/감시.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class Main {
    // 동 남,서,북
    static int[] dx = {0, 1, 0, -1};
    static int[] dy = {1, 0, -1, 0};
    static int n;
    static int m;
    static int[][] arr;
    static ArrayList<CCTV> cctvList = new ArrayList<>();
    static int ans = (int) 1e9;
    static StringBuilder sb = new StringBuilder();

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        arr = new int[n][m];

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < m; j++) {
                int num = Integer.parseInt(st.nextToken());
                arr[i][j] = num;

                if (num >= 1 && num <= 5) { // cctvList에 cctv 추가
                    cctvList.add(new CCTV(i, j, num));
                }
            }
        }
        dfs(0);

        sb.append(ans);
        System.out.println(sb);
    }

    static void dfs(int cnt) {
        if (cnt == cctvList.size()) {
            calculate();
            return;
        }

        CCTV cctv = cctvList.get(cnt);
        switch (cctv.num) {
            case (1):
                startCCTV1(cctv, cnt);
                break;
            case (2):
                startCCTV2(cctv, cnt);
                break;
            case (3):
                startCCTV3(cctv, cnt);
                break;
            case (4):
                startCCTV4(cctv, cnt);
                break;
            case (5):
                startCCTV5(cctv, cnt);
        }
    }

    static void startCCTV1(CCTV cctv, int cnt) {
        for (int i = 0; i < 4; i++) {
            go(cctv, i, false);
            dfs(cnt + 1);
            go(cctv, i, true);
        }
    }

    static void startCCTV2(CCTV cctv, int cnt) {
        for (int i = 0; i < 2; i++) {
            go(cctv, i, false);
            go(cctv, i + 2, false);
            dfs(cnt + 1);
            go(cctv, i, true);
            go(cctv, i + 2, true);
        }
    }

    private static void startCCTV3(CCTV cctv, int cnt) {
        for (int i = 0; i < 4; i++) {
            go(cctv, i, false);
            go(cctv, (i + 1) % 4, false);
            dfs(cnt + 1);
            go(cctv, i, true);
            go(cctv, (i + 1) % 4, true);
        }
    }

    static void startCCTV4(CCTV cctv, int cnt) {
        for (int i = 0; i < 4; i++) {
            go(cctv, i, false);
            go(cctv, (i + 3) % 4, false);
            go(cctv, (i + 1) % 4, false);
            dfs(cnt + 1);
            go(cctv, i, true);
            go(cctv, (i + 3) % 4, true);
            go(cctv, (i + 1) % 4, true);
        }
    }

    private static void startCCTV5(CCTV cctv, int cnt) {
        for (int i = 0; i < 4; i++) {
            go(cctv, i, false);
        }
        dfs(cnt + 1);
        for (int i = 0; i < 4; i++) {
            go(cctv, i, true);
        }
    }

    // remove : true -> 백트래킹을 하면서 감시 구역을 다시 원래대로 돌리는 경우
    // remove : false -> 감시구역 탐색
    static void go(CCTV cctv, int dir, boolean remove) {
        int nx = cctv.x;
        int ny = cctv.y;

        while (true) {
            nx += dx[dir];
            ny += dy[dir];

            if (nx < 0 || ny < 0 || nx >= n || ny >= m || arr[nx][ny] == 6) {
                break;
            }

            if (remove && arr[nx][ny] != 0) {  // 백트래킹을 해야하는 경우
                arr[nx][ny]--;
                continue;
            }

            arr[nx][ny]++; // CCTV가 감시하는 구역이면 ++
        }
    }

    static void calculate() {
        int cnt = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (arr[i][j] == 0) {
                    cnt++;
                }

                if (cnt >= ans) {
                    return;
                }
            }
        }
        ans = Math.min(ans, cnt);
    }

    static class CCTV {
        int x;
        int y;
        int num;

        CCTV(int x, int y, int num) {
            this.x = x;
            this.y = y;
            this.num = num;
        }
    }
}